src/components/ActionButton/StatefulActionButton.ts

```diff
@@ -0,0 +1,162 @@
+     * and the transition will not occur.
+     */
+    allowedTransitions?: IStatefulActionButtonTransition[];
+    /**
+     * The initial state to be used by the StatefulActionButton instance.
+     * If nullish or missing from the states array, a warning will be emitted,
+     * no ActionButton will be constructed and the element will be hidden
+     */
+    initalState: IStatefulActionButtonState;
+}
+
+/**
+ * An action button able to handle multiple states and their transitions.
+ */
+export class StatefulActionButton {
+    static ID = 'StatefulActionButton';
+    private currentState: IStatefulActionButtonState;
+    private innerActionButton: ActionButton;
+
+    constructor(public element: HTMLElement, public options: IStatefulActionButtonOptions, public bindings?: IResultsComponentBindings) {
+        if (!this.areOptionsValid()) {
+            return;
+        }
+        this.currentState = this.options.initalState;
+        this.currentState.onStateEntry?.apply(this);
+        this.innerActionButton = new ActionButton(element, { ...this.options.initalState, click: this.handleClick.bind(this) }, bindings);
+    }
+
+    /**
+     * Switch the state of the instance if the state and the transition between the current and new state are allowed.
+     * @param state a state to try to switch to
+     */
+    public switchTo(state: IStatefulActionButtonState) {
+        if (!this.isTransitionAllowed(state)) {
+            console.warn(
+                `State '${state.name}' is not allowed on this StatefulActionButton.\nEnsure to use the object references used at the instantiation`
+            );
+            return;
+        }
+        this.currentState.onStateExit?.apply(this);
+        state.onStateEntry?.apply(this);
+        this.innerActionButton.updateIcon(state.icon);
+        this.innerActionButton.updateTooltip(state.tooltip);
+        this.currentState = state;
+    }
+
+    /**
+     * Return the current state of the instance.
+     */
+    public getCurrentState() {
+        return this.currentState;
+    }
+
+    /**
+     * Check if the options given to the constructor are valid.
+     * If not, it will also display the appropriate warning.
+     */
+    private areOptionsValid(): boolean {
+        if (!this.options.states?.length) {
+            console.warn('The stateful action button cannot render if no states are defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (!this.options.initalState) {
+            console.warn('The stateful action button cannot render if the inital states is not defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (this.options.states.indexOf(this.options.initalState) < 0) {
+            console.warn('The stateful action button cannot render if the initial state is not in the list of states.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
```

Comment:
Maybe you should transform those 3 check into 3 little private function that returns a boolean

Reply:
No I disagree, that's already what's the method is doing by abstracting that from the constructor.
I don't like spaghetti code with a lot of functions etc. The function `areOptionsValid` does what it says and follow its comment, splitting the code would be splitting hair.

Reply:
It's not spaghetti code, it's well splitted functions

Reply:
I would agree if this was in a constructor, but I firmly believe that the net balance of splitting this private function in another three is negative.
I think the three ifs actually back-to-back showcase the nature of composite guard of the `isOptionsValid` way better than a `this.isOptionStateSet() | this.isOptionsInitialStateSet() | this.isOptionsInitialStateInOptionsStates()`.

I tried, I feel the code is less meaningful with this refactor, therefore it's not an improvement, but an hindrance and overengineering.

Reply:
I agree with Louis that having a private method for each condition is too much split. However, when I first saw the `areOptionsValid` method, I though about asking to split the state validations in a separate private method.

We would end with `areOptionsValid` looking something like this:

    private areOptionsValid(): boolean {
        return this.areStatesValid() && this.areTransitionsValid();
    }

But honestly, I wasn't sure it was worth splitting.